### ISSUE TITLE
Support compression on file import and export

### ISSUE THREAD

**Issue by matthiaskoenig** [opencobra/cobrapy#812]
In the discussion with @Midnighter it came up that there should be a generic support for compressed files for the various io modules. I.e. cobrapy should support the reading and writing of compressed files to the various formats (JSON, MAT, SBML, YAML).
To avoid code duplication there should be a single implementation of the compression support.

Compression becomes important for the genome-scale models which are very large (uncompressed).

Just as a note: The new SBML parser supports reading compressed files from paths, but not yet from file handles. But support for compression on writing is missing.

**Comment by Midnighter** [comment]
> Just as a note: The new SBML parser supports reading compressed files from paths, but not yet from file handles. But support for compression on writing is missing.

I suppose libsbml doesn't support writing to a Python file stream, i.e., it can either take a filename or create a string?

**Comment by matthiaskoenig** [comment]
Unfortunately not. Libsbml only supports reading or writing SBML to strings or paths. So the support for file handles and file streams will be implemented in python (via temporary files and reading the SBML strings via `read()`).

**Comment by akaviaLab** [comment]
Is this still valid? 
I'm asking because cobra/sbml.py seems to work well with gz and bz2, both reading and writing

**Comment by matthiaskoenig** [comment]
I implemented the compression support via libsbml for SBML. But there is no compression support for the other input formats.

**Comment by akaviaLab** [comment]
Hi
Would you like me to use python modules gzip and bz2 to implement
reading/writing of compressed files for the other formats?
It would be slightly different from libsbml, which does it internally, but
then the other formats could use them as well. I'm not sure it makes sense
for matlab, but I might implement it for completeness.

**Comment by matthiaskoenig** [comment]
Yes, this was basically the idea. Especially the JSON can be compressed very efficiently.

**Comment by akaviaLab** [comment]
Okay - if you can review and hopefully merge #1245, which allows I/O to use
Paths, I can build upon that to get all formats to deal with compressed
files.
Probably a helper function/file in the io directory, and all the formats
will call it.
